File: Week03/python/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        n = len(nums)
        def helper(i ,tmp):
            res.append(tmp)
            for j in (i, n):
                helper(j+1, tmp+[nums[j]])

        helper(0, [])
        return res

File: Week03/go/subsets.go
package main

func subsets(nums []int) [][]int {
	ans := make([][]int, 0)

	var backtrace func(pos, num int, cur []int)

	backtrace = func(pos, num int, cur []int) {
		if len(cur) == num {
			tmp := make([]int, len(cur))
			copy(tmp, cur)
			ans = append(ans, tmp)
			return
		}

		for i := pos; i < len(nums); i++ {
			cur = append(cur, nums[i])
			backtrace(i+1, num, cur)
			cur = cur[:len(cur)-1]
		}
	}

	for i := 0; i <= len(nums); i++ {
		cur := make([]int, 0, i)
		backtrace(0, i, cur)
	}

	return ans
}

